# 1. Importar librerías

In [62]:
#Librerías para manipulación de datos
import pandas as pd

#Librerías para visualización
import matplotlib.pyplot as plt
import seaborn as sns
from plotly.subplots import make_subplots
import plotly.express as px
import plotly.graph_objects as go

#Importar módulo de funciones
import sys
sys.path.insert(0, 'utils')  # Añadir el directorio 'utils' al PATH
import funciones
from importlib import reload
reload(funciones)  # Recarga el módulo

<module 'funciones' from 'c:\\recursos_humanos\\utils\\funciones.py'>

# 2. Importar datos

In [63]:
data = funciones.get_data('data_train')

In [64]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4410 entries, 0 to 4409
Data columns (total 30 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   InfoDate                 4410 non-null   dbdate 
 1   EmployeeID               4410 non-null   Int64  
 2   Age                      4410 non-null   Int64  
 3   DistanceFromHome         4410 non-null   Int64  
 4   MonthlyIncome            4410 non-null   Int64  
 5   NumCompaniesWorked       4410 non-null   float64
 6   PercentSalaryHike        4410 non-null   Int64  
 7   StockOptionLevel         4410 non-null   Int64  
 8   TrainingTimesLastYear    4410 non-null   Int64  
 9   YearsAtCompany           4410 non-null   Int64  
 10  YearsSinceLastPromotion  4410 non-null   Int64  
 11  YearsWithCurrManager     4410 non-null   Int64  
 12  TotalWorkingYears        4410 non-null   float64
 13  EnvironmentSatisfaction  4410 non-null   float64
 14  JobSatisfaction         

# 3. Análisis de Variables Categóricas

### 3.1. ¿Cuál fue el porcentaje de renuncias durante el año 2015?

In [65]:
funciones.univariado_torta(data, 'Attrition')

Attrition  Frecuencia Absoluta Frecuencia Relativa
0        No                 3760            85.261 %
1        Sí                  650            14.739 %

Aproximadamente el 15% de los colaboradores se retiraron en el año 2015, ya sea porque fueron despedidos o renunciaron.

### 3.2. ¿Cuál es el departamento que presentó el porcentaje más alto de retiros durante el año 2015?

In [66]:
funciones.analisisBivariado(data,['Attrition','Department'],'v','group')

Tabla de contingencia:


Department  Human Resources  Research & Development     Sales
Attrition                                                    
No                 0.036702                0.655585  0.307713
Sí                 0.078462                0.643077  0.278462

De los colaboradores que se retiraron, el 64.31% pertenecen al departamento de Investigación y Desarrollo. Además, de los colaboradores de Recursos Humanos, la proporción de colaboradores que se retiraron es el doble que la proporción de los que no se retiraron. En términos generales, el departamento con mayor número de retiros es Investigación y Desarrollo, pero hay que revisar Recursos Humanos porque el porcentaje de representación de los que se retiran es mayor que los que se quedan.

### 3.3. ¿Los retiros están relacionados con la frecuencia con la que los colaboradores deben viajar?

In [67]:
funciones.analisisBivariado(data, ['Attrition', 'BusinessTravel'], 'v','relative')

Tabla de contingencia:


BusinessTravel  Non-Travel  Travel_Frequently  Travel_Rarely
Attrition                                                   
No                0.110372           0.169681       0.719947
Sí                0.053846           0.296923       0.649231

De los colaboradores que se retiraron en el 2015, hubo más presencia de colaboradores que debían viajar frecuentemente. Además, de los colaboradores que debían viajar frecuentemente, la proporción de colaboradores que no se retiraron menor a la proporción de colaboradores que sí se retiraron, por lo que esta variable puede ser importante para predecir el retiro de un empleado.

### 3.4. ¿Qué influencia tiene el nivel de educación en el porcentaje de retiros?

In [68]:
funciones.analisisBivariado(data, ['Attrition', 'Education'], 'v', 'relative')

Tabla de contingencia:


Education  Bachelor  Below College   College    Doctor    Master
Attrition                                                       
No         0.389362       0.117021  0.186968  0.033245  0.273404
Sí         0.387692       0.107692  0.220000  0.029231  0.255385

No hay una influencia clara del nivel de educación con los retiros.

### 3.5. ¿Cuáles son los campos de estudio con más retiros?

In [69]:
funciones.analisisBivariado(data, ['Attrition', 'EducationField'], 'v', 'relative')

Tabla de contingencia:


EducationField  Human Resources  Life Sciences  Marketing   Medical     Other  \
Attrition                                                                       
No                     0.013564       0.411170   0.108511  0.314628  0.057979   
Sí                     0.046154       0.418462   0.106154  0.321538  0.043077   

EducationField  Technical Degree  
Attrition                         
No                      0.094149  
Sí                      0.064615

De los que se retiran, el porcentaje de colaboradores de Recursos Humanos es mayor en comparación con el porcentaje de colaboradores que no se retiraron siendod e Recursos Humanos.

### 3.6. Género

In [70]:
funciones.analisisBivariado(data,['Attrition','Gender'],"h","group")

Tabla de contingencia:


Gender       Female      Male
Attrition                    
No         0.404255  0.595745
Sí         0.375385  0.624615

No hay una dominancia de género en los retiros.

### 3.7. Rol del empleo

In [71]:
funciones.analisisBivariado(data,['Attrition','JobRole'],"v","relative")

Tabla de contingencia:


JobRole    Healthcare Representative  Human Resources  Laboratory Technician  \
Attrition                                                                      
No                          0.090957         0.035904               0.175798   
Sí                          0.078462         0.032308               0.178462   

JobRole     Manager  Manufacturing Director  Research Director  \
Attrition                                                        
No         0.071543                0.103457           0.050532   
Sí         0.056923                0.070769           0.076923   

JobRole    Research Scientist  Sales Executive  Sales Representative  
Attrition                                                             
No                   0.193617         0.220479              0.057713  
Sí                   0.227692         0.229231              0.049231

Solo el puesto de investigador científico tiene una mayor representación en los colaboradores que se retiran comparado con los que no.

### 3.8. Estado Civil

In [72]:
funciones.analisisBivariado(data,['Attrition','MaritalStatus'],"v","group")

Tabla de contingencia:


MaritalStatus  Divorced   Married    Single
Attrition                                  
No             0.236170  0.475798  0.288032
Sí             0.143077  0.353846  0.503077

Los colaboradores solteros tienen una mayor tendencia a retirarse de la empresa en comparación con los colaboradores casado o divorsiados.

# 4. Análisis de Satisfacción del Empleado

### 4.1. Satisfacción con el ambiente

In [73]:
funciones.analisisBivariado(data,['Attrition','EnvironmentSatisfaction'],"v","group")

Tabla de contingencia:


EnvironmentSatisfaction       1.0       2.0       3.0       4.0
Attrition                                                      
No                       0.172074  0.197606  0.318883  0.311436
Sí                       0.304615  0.173846  0.270769  0.250769

De los colaboradores que se retiraron de la empresa, la proporción que manifestaron una baja satisfacción con el ambiente es bastante mayor en comparación con los que no se retiraron. Por otro lado, los colaboradores que se quedaron presentaron una tendencia a manifestar una muy buena y excelente satisfacción con el ambiente.

### 4.2. Satisfacción con el empleo

In [74]:
funciones.analisisBivariado(data,['Attrition','JobSatisfaction'],"v","group")

Tabla de contingencia:


JobSatisfaction       1.0       2.0       3.0       4.0
Attrition                                              
No               0.180585  0.188298  0.299202  0.331915
Sí               0.278462  0.203077  0.304615  0.213846

Al igual que con la satisfacción con el entorno, la satisfacción con el empleo presenta calificaciones bajas por parte de los colaboradores que se retiraron de la empresa y mejores puntajes por parte de los que no se retiraron.

### 4.3. Balance entre vida y trabajo

In [75]:
funciones.analisisBivariado(data,['Attrition','WorkLifeBalance'],"v","group")

Tabla de contingencia:


WorkLifeBalance       1.0       2.0       3.0       4.0
Attrition                                              
No               0.044681  0.230851  0.623936  0.100532
Sí               0.109231  0.232308  0.541538  0.116923

EL balance entre vida y trabajo presenta calificaciones buenas tanto para colaboradores que se retiraron como para los que no. Aunque los que abandonaron tendieron un poco más a calificar esta variable como mala, hay una concentración alta en una buena calificación y poco en la mejor calificación.

# 5. Análisis de Desempeño del Empleado

### 5.1. Compromiso con el empleo

In [76]:
funciones.analisisBivariado(data,['Attrition','JobInvolvement'],"v","group")

Tabla de contingencia:


JobInvolvement         1         2         3         4
Attrition                                             
No              0.052660  0.255053  0.596809  0.095479
Sí              0.078462  0.255385  0.553846  0.112308

No hay una diferencia clara que permita concluir si esta variable influye en el retiro de los colaboradores.

In [77]:
funciones.analisisBivariado(data, ['retirementType', 'JobInvolvement'], 'v', 'relative')

Tabla de contingencia:


JobInvolvement         1         2         3         4
retirementType                                        
Fired           0.015385  0.353846  0.538462  0.092308
Resignation     0.085470  0.244444  0.555556  0.114530

No hay una indicación clara de que la calificación del compromiso con el empleo fuera una variable que influye con el retiro de los colaboradores y tampoco diferencia en tre los despedidos y los que renunciaron.

### 5.2. Calificación de Rendimiento

In [78]:
funciones.analisisBivariado(data,['Attrition','PerformanceRating'],"v","group")

Tabla de contingencia:


PerformanceRating         3         4
Attrition                            
No                 0.848404  0.151596
Sí                 0.833846  0.166154

El puntaje de rendimiento está casi que igualmente distribuido para ambos grupos, los que se retiraron y los que no.

In [79]:
funciones.analisisBivariado(data, ['retirementType', 'PerformanceRating'], 'v', 'relative')

Tabla de contingencia:


PerformanceRating         3         4
retirementType                       
Fired              0.784615  0.215385
Resignation        0.839316  0.160684

Se esperaría que los colaboradores despedidos presentara una mayor proporción en puntajes de rendimiento bajos, pero fue todo lo contrario.

Las variables de rendimiento, que son las que se esperaría que pudiesen explicar los despidos de colaboradores no lo están haciendo. Por lo que se opta por trabajar con algoritmos de cliasificación binaria, donde la salida sea si el colaborador se va o no.